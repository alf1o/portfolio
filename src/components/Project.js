import React from 'react';
import PropTypes from 'prop-types';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import Avatar from '@material-ui/core/Avatar';
import IconButton from '@material-ui/core/IconButton';
import OpenInNew from '@material-ui/icons/OpenInNew';
import Typography from '@material-ui/core/Typography';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import StarIcon from '@material-ui/icons/Star';
import { withStyles } from '@material-ui/core/styles';

const styles = theme => ({
  card: {
    width: '100vw',
    height: '100%',
    [theme.breakpoints.up('md')]: {
      width: '100%'
    }
  }
});

/**
  Render a single project.
*/
// TODO: open link icon is a `Link`
function Project({ classes, project }) {
  return (
    <Card
      className={classes.card}
    >
      <CardHeader
        avatar={
          <Avatar
            aria-label="project"
          >
            {project.icon}
          </Avatar>
        }
        action={
          <IconButton>
            <OpenInNew />
          </IconButton>
        }
        title={project.name}
      >
      </CardHeader>
      <CardContent>
        <Typography
          component="p"
        >
          {project.description}
        </Typography>
        <Typography
          component="p"
          variant="body2"
        >
          Features:
        </Typography>
        <List>
          {project.features.map(feature => (
            <ListItem key={feature}>
              <ListItemIcon>
                <StarIcon />
              </ListItemIcon>
              <ListItemText primary={feature} />
            </ListItem>
          ))}
        </List>
      </CardContent>
    </Card>
  );
}

Project.propTypes = {
  classes: PropTypes.object.isRequired,
  project: PropTypes.object.isRequired
};

export { Project };
export default withStyles(styles)(Project);
