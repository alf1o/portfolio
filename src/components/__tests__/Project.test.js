import React from 'react';
import Project from '../Project';
import { createShallow } from '@material-ui/core/test-utils';
import CardHeader from '@material-ui/core/CardHeader';
import ListItem from '@material-ui/core/ListItem';

describe('`Project`', () => {
  let props = {};
  let mountedPrj;
  const shallow = createShallow();

  function project() {
    if (!mountedPrj) mountedPrj = shallow(<Project {...props} />);
    return mountedPrj;
  }

  beforeEach(() => {
    props = {
      classes: {
        card: '',
        avatar: '',
        description: '',
        star: ''
      },
      project: {
        icon: 'T',
        url: 'test.com',
        name: 'test',
        description: 'test description',
        features: ['to', 'test', 'or', 'not', 'to', 'test']
      }
    };
    mountedPrj = undefined;
  });

  it('should render an `Avatar` icon which is `props.project.icon`', () => {
    const avt = project().dive().find(CardHeader).props().avatar;
    expect(avt.props.children).toBe(project().props().project.icon);
  });

  it('should render a `Button` which is a link to `project.url`', () => {
    const btn = project().dive().find(CardHeader).props().action;
    expect(btn.props.href).toBe(project().props().project.url);
  });

  it('should render a `CardHeader` with a `title` of `project.name`', () => {
    const title = project().dive().find(CardHeader).props().title;
    expect(title).toBe(project().props().project.name);
  });

  it('should render a `.description` with a text of `project.description`', () => {
    const desc = project().dive().find('.Project-description-3');
    expect(desc.props().children).toBe(project().props().project.description);
  });

  it('should render a `ListItem` for each `project.features`', () => {
    const ls = project().dive().find(ListItem);
    expect(ls.length).toBe(project().props().project.features.length);
  });

  it('should match snapshot', () => {
    expect(project()).toMatchSnapshot();
  });
});
