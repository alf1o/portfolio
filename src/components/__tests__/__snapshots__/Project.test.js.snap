// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Project\` should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Project)
    classes={
        Object {
            "avatar": "",
            "card": "",
            "description": "",
            "star": "",
          }
    }
    project={
        Object {
            "description": "test description",
            "features": Array [
              "to",
              "test",
              "or",
              "not",
              "to",
              "test",
            ],
            "icon": "T",
            "name": "test",
            "url": "test.com",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "classes": Object {
        "avatar": "Project-avatar-2",
        "card": "Project-card-1",
        "description": "Project-description-3",
        "star": "Project-star-4",
      },
      "project": Object {
        "description": "test description",
        "features": Array [
          "to",
          "test",
          "or",
          "not",
          "to",
          "test",
        ],
        "icon": "T",
        "name": "test",
        "url": "test.com",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "classes": Object {
          "avatar": "Project-avatar-2",
          "card": "Project-card-1",
          "description": "Project-description-3",
          "star": "Project-star-4",
        },
        "project": Object {
          "description": "test description",
          "features": Array [
            "to",
            "test",
            "or",
            "not",
            "to",
            "test",
          ],
          "icon": "T",
          "name": "test",
          "url": "test.com",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
